doc/rdup-tr.1.in
'\" t
.TH RDUP-TR 1 "27 Nov 2008" "@PACKAGE_VERSION@" "@PACKAGE_NAME@"
.SH NAME
@PACKAGE_NAME@-tr \- transform rdup output
.SH SYNOPSIS
.B @PACKAGE_NAME@-tr
[\fIOPTION\fR]...

.SH DESCRIPTION
Transform rdup output into \fIsomething else\fR. Where \fIsomething else\fR can
be a tar, cpio, pax archive or another rdup stream.

The rdup archive must be given on \fBrdup-tr\fR's standard input.

You can select multiple types of output (\-O flag), but you must be aware
that you may loose some information in formats other than \fBrdup\fR's own, see 
the table below.

You may also supply \fBrdup-tr\fR with only a list of pathnames, this can 
be selected with the \-L flag. 

The following table shows what happens with the output depending on the input. 
.BR
.TP 
.B 0
OK
.TP 
.B D
delete information is lost
.TP 
.B H
hardlink information is lost
.BR

.TS
L|L|CC.
    output	tar,cpio,pax	rdup
input	    
-------------	-------------	------
rdup	    D	0
filelist	    DH	H

.TE

.SH OPTIONS
.TP
.B \-L
Select list input format. Normally \fBrdup-tr\fR accepts rdup output,
with this option you can give it a list of path names.
Note: with list input \fBrdup-tr\fR will `stat()` each file, so this can not
be used in remote back ups.
.TP
.B \-O
Output format. This can be 'tar', 'cpio', 'pax' or 'rdup'. It
defaults to 'rdup'. 
.TP
.B \-X \fIkey\fR
Read the encryption key from the file key and encrypt all paths
with Blowfish and this key and iv. After the encryption the binary data is
converted into ASCII using an URL safe (Section 4 of RFC 3548)
version of base64 encode.

The encryption key must be on the first line and the key size
must be 16 and 8 bytes for the iv, so 24 in total.
.TP
.B \-Y \fIkey\fR
Read the decryption key from the file key and decrypt all paths
with Blowfish and this key. Before the encryption the paths
are converted to binary by using an URL safe version of base64 decode.
.TP
.B \-c
Force output to the tty. Normally \fBrdup-tr\fR wants to see it's
output redirected.
.TP
.B \-v
Be more verbose. Print the processed file names to standard error.
.TP
.B \-V
Print rdup-tr's version.
.TP
.B \-h
A short help.

.SH EXAMPLES
The following is possible

.RS
 rdup -P gzip -P "mcrypt -f KEY -c" /dev/null /home | \\
 rdup-tr -O tar -X<(echo secret) | gzip > \\
 my-home-zipped-crypted-pathcrypted-tar.gz
.RE

That is: all files under /home are gzipped and encrypted
on a \fIper\fR file basis (first line). Further more, all pathnames are
Blowfish encrypted (second line) with the key 'secret'.
This is put in a tar file, which is then
compressed, resulting in the final output (final line).

Creating a compressed and encrypted tar archive out of a full
rdup dump might be done as follows

.RS
 rdup -P gzip -P "mcrypt -f KEY -c" /dev/null /home | \\
 rdup-tr -O tar > my-home-zipped-and-crypted.tar
.RE

Or even pack and unpack it on the fly

.RS
 rdup -P gzip -P "mcrypt -fKEY -c" /dev/null /home | rdup-tr -Otar | \\
 ssh user@remotehost tar xvCf /tmp -
.RE

Or encryption with openssl

.RS
rdup -P "openssl enc -e -des-cbc -k secret" /dev/null /home
.RE

Or, compressing with gzip, encrypting with openssl and then compressing
the entire archive yet again

.RS
 rdup -P gzip -P "openssl enc -e -des-cbc -k secret" /dev/null /home | \\
 gzip > my_compressed_encrypted_rdup_archive.gz
.RE

.SH EXIT CODE
\fBrdup-tr\fR return a zero exit code on success, otherwise 1 is returned.

.SH AUTHOR
Written by Miek Gieben. 

.SH REPORTING BUGS
Report bugs to <miek@miek.nl>. 

.SH SEE ALSO
http:/www.miek.nl/projects/rdup/ is the main site of rdup. Also see rdup(1),
rdup-up(1) and rdup-backups(7).

.SH COPYRIGHT
Copyright (C) 2005-2010 Miek Gieben. This is free software. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.
.PP
Licensed under the GPL version 3. See the file LICENSE in the source distribution
of rdup.


ChangeLog
xx Yyy 2013: 1.1.15 Miek Gieben <miek@miek.nl>
  * [rdup] Change -P so any arguments are executed with 'sh -c', 
    this removes the 7 option limit (Alexander Neumann)
  * [rdup-tr] Fix crashes with filename encryption and symlinks (reported by
    Jakob Fink)

19 Jan 2013: 1.1.14 Miek Gieben <miek@miek.nl>
   * [rdup] check for the existence of ._rdup_. files and parse its contents
     to restore the actual files' ownership (Miek Gieben)
   * [rdup-up] save user/group info in ._rdup_. files when chown(1)ing has failed.
     Useful when backing up via sshfs. See rdup(1) "Non-root backups" (Miek Gieben)
   * [rdup-up] add -u flag to suppress creating these ._rdup_. files. 
     (useful when restoring) (Miek Gieben)
   * [rdup] add -u flag to suppress interpreting these ._rdup_. files. (Miek Gieben)
   * [rdup-tr] compile fix when compiling with gcc 4.7 (reported by D. L.)
   * [rdup-tr] honor the TMPDIR environment variable (Miek Gieben)
   * [rdup] Don't look at the device ID when deciding if a file needs to be backed up 
     (See https://github.com/miekg/rdup/issues/3) 


18 Feb 2012: 1.1.13 Miek Gieben <rdup@miek.nl>
   * Make rdup-tr compile when libnettle was not found (Miek Gieben)
   * Fix for very large inodes and filesizes (Kojedzinszky Richard)
   * Fix tar/pax output when the filesize of files changes (rdup -Pgzip for
     instance) (Reported by: Alejandro Pulver). This created files padded
     with zero bytes.


21 Jan 2012: 1.1.12 Miek Gieben <rdup@miek.nl>
   * Test suite fixed
   * Fix Debian bug 627617


18 Feb 2011. 1.1.11 Miek Gieben <rdup@miek.nl>
   * Polish
   * rdup-up: added -q option. Be quiet when chown
     fails even when root (handy for backups on sshfs)


05 Dec 2010: 1.1.10 Miek Gieben <rdup@miek.nl>
   * One memleak fix removed (Tom Hendrikx)


04 Dec 2010: 1.1.9 Miek Gieben <rdup@miek.nl>
   * Fix whooping amount of memleaks (rdup-tr, rdup-up) (Crispin Boylan)


24 Oct 2010: 1.1.8 Miek Gieben <rdup@miek.nl>
   * rdup: Check for two more I/O errors when reading files
   * rdup: Permission fix for AIX (Crispin Boylan)
   * autoconf: Fix dirfd detection on mac OS X (still does not work)


21 Jun 2010: 1.1.7 Miek Gieben <rdup@miek.nl>
    * rdup-up: also create named pipes (fifo)


18 May 2010: 1.1.6 Miek Gieben <rdup@miek.nl>
    * rdup-simple: cleanups + portability


09 Apr 2010: 1.1.5 Miek Gieben <rdup@miek.nl>
    * compile fix


04 Apr 2010: 1.1.4 Miek Gieben <rdup@miek.nl>
    * re-touch timestamp after the backup is made when -a is used,
      otherwise it is done before the backup
    * spring cleaning, remove/clean BUGBUGs and some dead code and
      polish the documentation a bit
    * revert changes in mkpath.c to make FreeBSD happy
    * fix hardlinks in tar/pax output (they must come last)


22 Mar 2010: 1.1.3 Miek Gieben <rdup@miek.nl>
    * rdup: fix reset atime (-a)
    * rdup-simple: add -a (reset atime flag)
    * documentation bugfixes


14 Mar 2010: 1.1.2 Miek Gieben <rdup@miek.nl>
    * rdup: remove sig_atomic_t (build fix for ubuntu 8.04)
     (reported by Andrew Ittner)
    * rdup-up: fixed nasty bug in dir.c dir_parent()
     (leave caller's string alone)
    * rdup-up: rm.c: also use dir_parent()
    * tests: a 'fail'ing gpg test is now 'unresolved'
    * Added contrib/ directory and added prune.rb script
      (prune.rb provided by Matthias-Christian Ott)


2 Mar 2010: 1.1.1 Miek Gieben <rdup@miek.nl>
    * tests: Skip gpg test when gpg is not found (Tom)
    * rdup: Rename strmode() to rdup_strmode for FreeBSD (Charlie Kester)
    * rdup: Dont descend when toplevel directory is a symlink (same behavoir
      as Gnu tar) (Reported by Charlie Kester)
    * rdup: reap our childeren with waitpid() (Oliver Dain)
    * build: Building fix on rpm distros (SuSE/Fedora) (Sven Hartrumpf)
    * build: add --enable-debug to build debug build


24 Jan 2010: 1.1.0 Miek Gieben <rdup@miek.nl>
    * Release!
    * Tweaks and merges from the master branch
    * Remember: installs in /usr/local


22 Dec 2009: 1.1.0-rc1 Miek Gieben <rdup@miek.nl>
    * First release of the 1.1.0 branch
    * format change: %U, %G and %t added
    * default install path: /usr/local/
    * rdup: -a restore atime when reading files
    * rdup: -P is moved from rdup-tr 
    Todo:
    * documentation
    * more testing


14 Dec 2009: 1.1.0-dev1 Miek <rdup@miek.nl>
    * rdup added:
	- -a: reset atime
    * rdup-up:
	- set mtime
    * revert back to libnettle
	- and back to AES encryption


09 Aug 2009: 1.1.0-dev0 Miek <rdup@miek.nl>
    * Format change: add username, groupname and mtime
      Always output the files' contents
      %U username
      %G groupname
      %t is mtime of the fs object
    * Adapt all utils to this format change
	- rdup now handles -P itself
	- rdup is the only tool that accesses the files
	- rdup-tr does not do -P anymore
	- rdup-tr only processes from stdin to stdout
	- rdup-up only processes stdin
    * Loads of code cleanups
    * Use OpenSSL in favor of libnettle
	- AES encryption is replaced by blowfish
    * Change default install path /usr/local (Sebastian Kügler)


rdup-tr.c
/*
 * Copyright (c) 2009 - 2011 Miek Gieben
 * License: GPLv3(+), see LICENSE for details
 * rdup-tr -- rdup translate, transform an
 * rdup filelist to an tar/cpio archive with
 * per file compression and/or encryption or whatever
 */

/* stdio is only used for errors
 * the rest is all read, write and pipe
 */

#include "rdup-tr.h"
#include "protocol.h"
#include "io.h"

char *PROGNAME = "rdup-tr";
/* options */
char *template;
gboolean opt_tty = FALSE;	/* force write to tty */
#ifdef HAVE_LIBNETTLE
gchar *opt_crypt_key = NULL;	/* encryption key */
gchar *opt_decrypt_key = NULL;	/* decryption key */
struct aes_ctx *aes_ctx = NULL;
#endif				/* HAVE_LIBNETTLE */
gint opt_verbose = 0;		/* be more verbose */
gint opt_output = O_RDUP;	/* default output */
gint opt_input = I_RDUP;	/* default intput */

int sig = 0;
char *o_fmt[] = { "", "tar", "cpio", "pax", "rdup" };	/* O_NONE, O_TAR, O_CPIO, O_PAX, O_RDUP */

extern int opterr;

int opterr = 0;

/* common.c */
struct rdup *entry_dup(struct rdup *f);
void entry_free(struct rdup *f);

#ifdef HAVE_LIBNETTLE
/* encrypt an rdup_entry (just the path of course) */
static struct rdup *crypt_entry(struct rdup *e, GHashTable * tr)
{
	gchar *crypt, *dest;
	if (!(crypt = crypt_path(aes_ctx, e->f_name, tr))) {
		msg(_("Failed to encrypt path `%s\'"), e->f_name);
		return NULL;
	}

	e->f_name = crypt;
	e->f_name_size = strlen(crypt);

	/* links are special */
	if (S_ISLNK(e->f_mode) || e->f_lnk == 1) {
		dest = crypt_path(aes_ctx, e->f_target, tr);
		e->f_target = dest;
		e->f_size = strlen(crypt);	/* use crypt here */
	}
	return e;
}

/* decrypt an rdup_entry */
static struct rdup *decrypt_entry(struct rdup *e, GHashTable * tr)
{
	gchar *plain, *dest;

	if (!(plain = decrypt_path(aes_ctx, e->f_name, tr))) {
		msg(_("Failed to decrypt path `%s\'"), e->f_name);
		return NULL;
	}

	e->f_name = plain;
	e->f_name_size = strlen(plain);

	/* links are special */
	if (S_ISLNK(e->f_mode) || e->f_lnk == 1) {
		dest = decrypt_path(aes_ctx, e->f_target, tr);
		g_free(e->f_target);
		e->f_target = dest;
		e->f_size = strlen(plain);
	}
	return e;
}
#endif				/* HAVE_LIBNETTLE */

void tmpclean(int opt_output, int f, char *name)
{
	if (opt_output != O_RDUP) {
		if (f != -1)
			close(f);
		unlink(name);
		g_free(name);
	}
}

/* read filenames from stdin, put them through
 * the childeren, collect the output from the last
 * child and create the archive on stdout
 */
static void stdin2archive(void)
{
	char *buf, *fbuf, *readbuf, *n, *pathbuf;
	char *tmpname = NULL;
	char delim;
	size_t i, line, pathsize;
	ssize_t bytes;
	int j, k;
	int tmpfd = -1;
	struct archive *archive;
	struct archive_entry *entry;
	struct stat *s;
	struct rdup *rdup_entry = NULL;
	GSList *hlinks = NULL;
	GSList *hl = NULL;
#ifdef HAVE_LIBNETTLE
	GHashTable *trhash;	/* look up for encrypted/decrypted strs */
	trhash = g_hash_table_new(g_str_hash, g_str_equal);
#endif				/* HAVE_LIBNETTLE */

	delim = '\n';
	i = BUFSIZE;
	buf = g_malloc(BUFSIZE + 1);
	fbuf = g_malloc(BUFSIZE + 1);
	readbuf = g_malloc(BUFSIZE + 1);
	pathbuf = g_malloc(BUFSIZE + 1);
	j = ARCHIVE_OK;
	entry = NULL;
	line = 0;

	if (opt_output == O_RDUP) {
		archive = NULL;
	} else {
		if ((archive = archive_write_new()) == NULL) {
			msg(_("Failed to create archive"));
			exit(EXIT_FAILURE);
		}

		switch (opt_output) {
		case O_TAR:
			j = archive_write_set_format_ustar(archive);
			break;
		case O_CPIO:
			j = archive_write_set_format_cpio(archive);
			break;
		case O_PAX:
			j = archive_write_set_format_pax(archive);
			break;
		case O_RDUP:
			j = ARCHIVE_OK;
			break;
		}

		if (j != ARCHIVE_OK) {
			msg(_("Failed to set archive type to %s"),
			    o_fmt[opt_output]);
			exit(EXIT_FAILURE);
		} else {
			archive_write_open_fd(archive, 1);
		}
	}

	char *tmpenv = getenv("TMPDIR");
	if (!tmpenv) {
		tmpenv = "/tmp";
	}

	while ((rdup_getdelim(&buf, &i, delim, stdin)) != -1) {
		if (opt_output != O_RDUP) {
			tmpname =
			    g_strdup_printf("%s/%s", tmpenv, "rdup-XXXXXX");
			tmpfd = mkstemp(tmpname);
			if (tmpfd == -1) {
				msg(_
				    ("Failure to open temporary file `%s\': %s\n"),
				    tmpname, strerror(errno));
				exit(EXIT_FAILURE);
			}
		}
		line++;
		n = strrchr(buf, '\n');
		if (n)
			*n = '\0';

		if (sig != 0) {
			tmpclean(opt_output, tmpfd, tmpname);
			signal_abort(sig);
		}

		if (!(rdup_entry = parse_entry(buf, line))) {
			/* msgs from entry.c */
			tmpclean(opt_output, tmpfd, tmpname);
			exit(EXIT_FAILURE);
		}

		/* we have a valid entry, read the filename */
		pathsize =
		    fread(pathbuf, sizeof(char), rdup_entry->f_name_size,
			  stdin);

		if (pathsize != rdup_entry->f_name_size) {
			msg(_
			    ("Reported name size (%zd) does not match actual name size (%zd)"),
			    rdup_entry->f_name_size, pathsize);
			tmpclean(opt_output, tmpfd, tmpname);
			exit(EXIT_FAILURE);
		}
		pathbuf[pathsize] = '\0';
		if (pathbuf[0] != '/') {
			msg(_("Pathname does not start with /: `%s\'"),
			    pathbuf);
			tmpclean(opt_output, tmpfd, tmpname);
			exit(EXIT_FAILURE);
		}

		g_free(rdup_entry->f_name);
		rdup_entry->f_name = pathbuf;

		/* extract target from rdup_entry */
		if (S_ISLNK(rdup_entry->f_mode) || rdup_entry->f_lnk) {
			// filesize is spot where to cut and set new size
			rdup_entry->f_name[rdup_entry->f_size] = '\0';
			rdup_entry->f_name_size = strlen(rdup_entry->f_name);
			g_free(rdup_entry->f_target);
			rdup_entry->f_target = rdup_entry->f_name +
			    rdup_entry->f_size + 4;
		} else {
			rdup_entry->f_target = NULL;
		}

		if (opt_verbose > 0) {
			if (S_ISLNK(rdup_entry->f_mode) || rdup_entry->f_lnk) {
				fprintf(stderr, "%s -> %s\n",
					rdup_entry->f_name,
					rdup_entry->f_target);
			} else {
				fprintf(stderr, "%s\n", rdup_entry->f_name);
			}
		}

		if (sig != 0) {
			tmpclean(opt_output, tmpfd, tmpname);
			signal_abort(sig);
		}
#ifdef HAVE_LIBNETTLE
		if (opt_crypt_key)
			rdup_entry = crypt_entry(rdup_entry, trhash);
		if (opt_decrypt_key)
			rdup_entry = decrypt_entry(rdup_entry, trhash);

		if (!rdup_entry) {
			tmpclean(opt_output, tmpfd, tmpname);
			exit(EXIT_FAILURE);	/* encryption problem */
		}
#endif				/* HAVE_LIBNETTLE */

		if (rdup_entry->plusmin == MINUS) {
			if (opt_output == O_RDUP) {
				(void)rdup_write_header(rdup_entry);
				goto not_s_isreg;
			}
			g_free(rdup_entry);
			continue;
		}

		if (opt_output != O_RDUP) {
			if (rdup_entry->f_lnk == 1) {
				/* hardlinks must come last */
				hlinks =
				    g_slist_append(hlinks,
						   entry_dup(rdup_entry));
				g_free(rdup_entry);
				tmpclean(opt_output, tmpfd, tmpname);
				continue;
			}

			s = stat_from_rdup(rdup_entry);
			entry = archive_entry_new();
			archive_entry_copy_stat(entry, s);
			archive_entry_copy_pathname(entry, rdup_entry->f_name);

			if (S_ISLNK(rdup_entry->f_mode))
				archive_entry_copy_symlink(entry,
							   rdup_entry->
							   f_target);
			g_free(s);
		}

		if (opt_output == O_RDUP)
			(void)rdup_write_header(rdup_entry);

		/* bail out for non regular files */
		if (!S_ISREG(rdup_entry->f_mode) || rdup_entry->f_lnk == 1) {
			// We haven't written a header yet for != O_RDUP
			if (opt_output != O_RDUP) {
				archive_write_header(archive, entry);
			}
			goto not_s_isreg;
		}

		/* regular files */
		ssize_t totalbytes = 0;
		while ((bytes = block_in_header(stdin)) > 0) {
			totalbytes += bytes;
			if (block_in(stdin, bytes, fbuf) == -1) {
				msg(_("Failure to read from stdin: %s"),
				    strerror(errno));
				tmpclean(opt_output, tmpfd, tmpname);
				exit(EXIT_FAILURE);
			}

			if (sig != 0) {
				tmpclean(opt_output, tmpfd, tmpname);
				signal_abort(sig);
			}

			if (opt_output == O_RDUP)
				(void)rdup_write_data(rdup_entry, fbuf, bytes);

			if (opt_output != O_RDUP) {
				// Write to temp file
				if (write(tmpfd, fbuf, bytes) != bytes) {
					msg(_
					    ("Failure to write to temporary file `%s\': %s"),
					    tmpname, strerror(errno));
					tmpclean(opt_output, tmpfd, tmpname);
					exit(EXIT_FAILURE);
				}
			}
		}
		/* final block for rdup output */
		if (opt_output == O_RDUP)
			block_out_header(NULL, 0, 1);

		if (opt_output != O_RDUP) {
			// For output that is not O_RDUP we saved the data in a temporary file.
			// Read back this data and output the archive with the correct size.
			archive_entry_set_size(entry, totalbytes);
			archive_write_header(archive, entry);
			if (lseek(tmpfd, 0, SEEK_SET) == -1) {
				msg(_
				    ("Failure to rewind temporary file `%s\': %s"),
				    tmpname, strerror(errno));
				tmpclean(opt_output, tmpfd, tmpname);
				exit(EXIT_FAILURE);
			}

			k = read(tmpfd, fbuf, BUFSIZE);
			while (k > 0) {
				archive_write_data(archive, fbuf, k);
				k = read(tmpfd, fbuf, BUFSIZE);
				if (sig != 0) {
					tmpclean(opt_output, tmpfd, tmpname);
					signal_abort(sig);
				}
			}
		}

 not_s_isreg:
		if (opt_output != O_RDUP)
			archive_entry_free(entry);

		g_free(rdup_entry->f_user);
		g_free(rdup_entry->f_group);
		g_free(rdup_entry);
		tmpclean(opt_output, tmpfd, tmpname);
	}
	/* output hardlinks -- if any, is zero for O_RDUP */
	for (hl = g_slist_nth(hlinks, 0); hl; hl = hl->next) {
		s = stat_from_rdup((struct rdup *)hl->data);
		entry = archive_entry_new();
		archive_entry_copy_stat(entry, s);
		archive_entry_copy_pathname(entry,
					    ((struct rdup *)hl->data)->f_name);
		archive_entry_copy_hardlink(entry,
					    ((struct rdup *)hl->data)->
					    f_target);
		archive_write_header(archive, entry);
		archive_entry_free(entry);
		g_free(s);
	}

	if (opt_output != O_RDUP) {
		archive_write_close(archive);
		archive_write_free(archive);
	}
	g_free(readbuf);
	g_free(buf);
}

int main(int argc, char **argv)
{
	struct sigaction sa;
	char pwd[BUFSIZE + 1];
	int c;

#ifdef ENABLE_NLS
	if (!setlocale(LC_MESSAGES, ""))
		msg(_("Locale could not be set"));
	bindtextdomain(PACKAGE_NAME, LOCALEROOTDIR);
	(void)textdomain(PACKAGE_NAME);
#endif				/* ENABLE_NLS */

	/* setup our signal handling */
	sa.sa_flags = 0;
	sigfillset(&sa.sa_mask);

	sa.sa_handler = got_sig;
	sigaction(SIGPIPE, &sa, NULL);
	sigaction(SIGINT, &sa, NULL);

	if (((getuid() != geteuid()) || (getgid() != getegid()))) {
		msg(_("Will not run suid/sgid for safety reasons"));
		exit(EXIT_FAILURE);
	}

	if (!getcwd(pwd, BUFSIZE)) {
		msg(_("Could not get current working directory"));
		exit(EXIT_FAILURE);
	}

	for (c = 0; c < argc; c++) {
		if (strlen(argv[c]) > BUFSIZE) {
			msg(_("Argument length overrun"));
			exit(EXIT_FAILURE);
		}
	}

	while ((c = getopt(argc, argv, "cP:O:t:LhVvX:Y:")) != -1) {
		switch (c) {
		case 'c':
			opt_tty = TRUE;
			break;
		case 'v':
			opt_verbose++;
			break;
		case 'L':
			opt_input = I_LIST;
			break;
		case 'P':
			msg(_("Functionality moved to rdup"));
			break;
		case 'O':
			opt_output = O_NONE;

			if (strcmp(optarg, o_fmt[O_TAR]) == 0)
				opt_output = O_TAR;
			if (strcmp(optarg, o_fmt[O_CPIO]) == 0)
				opt_output = O_CPIO;
			if (strcmp(optarg, o_fmt[O_PAX]) == 0)
				opt_output = O_PAX;
			if (strcmp(optarg, o_fmt[O_RDUP]) == 0)
				opt_output = O_RDUP;

			if (opt_output == O_NONE) {
				msg(_("Invalid output format: `%s\'"), optarg);
				exit(EXIT_FAILURE);
			}
			break;
		case 'X':
#ifdef HAVE_LIBNETTLE
			if (opt_decrypt_key) {
				msg(_
				    ("Will not do both encryption and decryption"));
				exit(EXIT_FAILURE);
			}
			if (!(opt_crypt_key = crypt_key(optarg)))
				exit(EXIT_FAILURE);

			aes_ctx = crypt_init(opt_crypt_key, TRUE);
			if (!aes_ctx)
				exit(EXIT_FAILURE);
#else
			msg(_("Compiled without encryption, can not encrypt"));
			exit(EXIT_FAILURE);
#endif				/* HAVE_LIBNETTLE */
			break;
		case 'Y':
#ifdef HAVE_LIBNETTLE
			if (opt_crypt_key) {
				msg(_
				    ("Can not do both encryption and decryption"));
				exit(EXIT_FAILURE);
			}
			if (!(opt_decrypt_key = crypt_key(optarg)))
				exit(EXIT_FAILURE);

			aes_ctx = crypt_init(opt_decrypt_key, FALSE);
			if (!aes_ctx)
				exit(EXIT_FAILURE);
#else
			msg(_("Compiled without encryption, can not decrypt"));
			exit(EXIT_FAILURE);
#endif				/* HAVE_LIBNETTLE */
			break;
		case 'h':
			usage_tr(stdout);
			exit(EXIT_SUCCESS);
		case 'V':
#ifdef DEBUG
			fprintf(stdout, "%s %s (with --enable-debug)\n",
				PROGNAME, VERSION);
#else
			fprintf(stdout, "%s %s\n", PROGNAME, VERSION);
#endif				/* DEBUG */
			exit(EXIT_SUCCESS);
		default:
			msg(_("Unknown option seen `%c\'"), optopt);
			exit(EXIT_FAILURE);
		}
	}
	argc -= optind;
	argv += optind;

	if (!opt_tty && isatty(1) == 1) {
		msg(_("Will not write to a tty"));
		exit(EXIT_FAILURE);
	}

	/* read stdin and (re)make an archive */
	stdin2archive();
	exit(EXIT_SUCCESS);
}
